bin/coqtop.ml

```diff
@@ -0,0 +1,131 @@
+    Arg.(value & opt string "coqtop" & info [ "toplevel" ] ~docv:"CMD" ~doc)
+  and+ coq_file_arg =
+    Arg.(required & pos 0 (some string) None (Arg.info [] ~docv:"COQFILE"))
+  and+ extra_args =
+    Arg.(value & pos_right 0 string [] (Arg.info [] ~docv:"ARGS"))
+  in
+  let config = Common.init common in
+  let coqtop, argv, env =
+    Scheduler.go ~common ~config (fun () ->
+        let open Fiber.O in
+        let* setup = Import.Main.setup () in
+        let* setup = Memo.Build.run setup in
+        let sctx = Import.Main.find_scontext_exn setup ~name:context in
+        let context = Dune_rules.Super_context.context sctx in
+        (* Try to compute a relative path if we got an absolute path. *)
+        let coq_file_arg =
+          let cwd = Path.external_ (Path.External.initial_cwd) in
+          let file =
+            (* Best-effort symbolic link unfolding. *)
+            let file = Fpath.follow_symlinks coq_file_arg in
+            Option.value file ~default:coq_file_arg
+          in
+          let file = Path.of_filename_relative_to_initial_cwd file in
+          let cwd = Path.to_string cwd in
+          let file = Path.to_string file in
+          match String.drop_prefix ~prefix:(cwd ^ "/") file with
+          | None -> coq_file_arg
+          | Some s -> s
+        in
+        let coq_file_build =
+          let p = Common.prefix_target common coq_file_arg in
+          Path.Build.relative context.build_dir p
+        in
+        let dir =
+          let dir = Filename.dirname coq_file_arg in
+          let p = Common.prefix_target common dir in
+          Path.Build.relative context.build_dir p
+        in
+        let* (coqtop, args) =
+          let open Memo.Build.O in
+          Build_system.run_exn (fun () ->
+              let* dc = Dune_rules.Dir_contents.get sctx ~dir in
+              let* coq_src = Dune_rules.Dir_contents.coq dc in
+              let coq_module =
+                let source = coq_file_build in
+                match Dune_rules.Coq_sources.find_module ~source coq_src with
+                | Some m -> snd m
+                | None ->
+                    let hints =
+                      [ Pp.textf "is the file part of a stanza?"
+                      ; Pp.textf "has the file been written to disk?" ]
+                    in
+                    User_error.raise ~hints
+                      [ Pp.textf "cannot find file: %s" coq_file_arg ]
+              in
+              let stanzas =
+                let name = Dune_rules.Coq_module.name coq_module in
+                Dune_rules.Coq_sources.lookup_module coq_src name
+              in
+              let* (args, boot_type) =
+                match stanzas with
+                | [] ->
+                    User_error.raise
+                      [ Pp.textf "file not part of any stanza: %s" coq_file_arg ]
+                | `Theory theory :: stanzas ->
+                    if stanzas <> [] then assert false; (* FIXME *)
```

Comment:
What should we do here? (And similarly below.)

Reply:
This was silly, the mistake was that the domain of the map should be Coq modules, not Coq module names...